hbase-it/src/test/java/org/apache/hadoop/hbase/KubernetesClusterManager.java

```diff
@@ -0,0 +1,258 @@
+ * will handle starting replacement processes after a destructive action. Requires that the
+ * configuration specify a Kubernetes namespace.
+ * </p>
+ * Note that the k8s java client is a bit dodgy unable to read responses when we call delete (i.e.
+ * kill) and sometimes when checking isRunning; makes for noisy logs and sometimes the operations
+ * overrun each other but generally succeed.
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.CONFIG)
+public class KubernetesClusterManager extends Configured implements ClusterManager {
+  private static final Logger LOG = LoggerFactory.getLogger(KubernetesClusterManager.class);
+
+  private static final String NAMESPACE_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".namespace";
+  private static final String ZOOKEEPER_ROLE_SELECTOR_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".zookeeper_role";
+  private static final String NAMENODE_ROLE_SELECTOR_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".namenode_role";
+  private static final String JOURNALNODE_ROLE_SELECTOR_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".journalnode_role";
+  private static final String DATANODE_ROLE_SELECTOR_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".datanode_role";
+  private static final String MASTER_ROLE_SELECTOR_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".master_role";
+  private static final String REGIONSERVER_ROLE_SELECTOR_CONF_KEY =
+    KubernetesClusterManager.class.getCanonicalName() + ".regionserver_role";
+  private static final Set<ServiceType> SUPPORTED_SERVICE_TYPES = buildSupportedServiceTypesSet();
+
+  private CoreV1Api api;
+  private String namespace;
+
+  private static Set<ServiceType> buildSupportedServiceTypesSet() {
+    final Set<ServiceType> set = new HashSet<>();
+    set.add(ServiceType.ZOOKEEPER_SERVER);
+    set.add(ServiceType.HADOOP_NAMENODE);
+    set.add(ServiceType.HADOOP_JOURNALNODE);
+    set.add(ServiceType.HADOOP_DATANODE);
+    set.add(ServiceType.HBASE_MASTER);
+    set.add(ServiceType.HBASE_REGIONSERVER);
+    return Collections.unmodifiableSet(set);
+  }
+
+  @Override
+  public void setConf(Configuration configuration) {
+    // This is usually called from the constructor (The call to super goes to
+    // Configured which does a setConf in its construction) but the configuration
+    // null at that time. After construction, setConf is called again.
+    // Assume single-thread doing ClusterManager setup.
+    if (configuration == null) {
+      LOG.debug("Skipping because provided configuration=null");
+      return;
+    }
+    if (getConf() != null) {
+      LOG.debug("Skipping because configuration already set, getConf={}", getConf());
+      return;
+    }
+    super.setConf(configuration);
+    this.namespace = configuration.get(NAMESPACE_CONF_KEY);
```

Comment:
Use the two-arg `get()` method to give a default value?

Reply:
What default namespace would you propose? I guess, `default` ?

Reply:
That would be the sane k8s thing to do. If you have a namespace you use typically use (e.g. "hbase-testing"), that's also cool. Anything to avoid the NPE.